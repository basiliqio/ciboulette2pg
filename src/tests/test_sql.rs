use super::*;
pub async fn snapshot_table(
    transaction: &mut sqlx::Transaction<'_, sqlx::Postgres>,
    name: &str,
    tables: &[&str],
) {
    let mut map = serde_json::Map::new();

    for table in tables.iter() {
        let snap: (Option<serde_json::Value>,) = sqlx::query_as(
            format!(
				"SELECT ARRAY_TO_JSON(ARRAY_AGG(result)) AS data FROM (SELECT * FROM \"{}\") result;",
				table
			)
            .as_str(),
        )
        .fetch_one(&mut *transaction)
        .await
        .expect("to result the database snapshot");
        match snap.0 {
            Some(x) => map.insert(table.to_string(), x),
            None => map.insert(table.to_string(), serde_json::Value::Null),
        };
    }

    insta::assert_json_snapshot!(name,
        serde_json::Value::Object(map),
    {
        ".*.*.id" => insta::dynamic_redaction(check_uuid),
        ".*.*.people_id" => insta::dynamic_redaction(check_uuid),
        ".*.*.article_id" => insta::dynamic_redaction(check_uuid),
        ".*.*.favorite_color" => insta::dynamic_redaction(|value, _path| {
            match value
            .as_str()
            {
                Some(x) => {
                    assert_eq!(x
                        .chars()
                        .filter(|&c| c == '-')
                        .count(),
                        4
                    );
                    "[favorite_color_uuid]"
                },
                    None => "[favorite_color_null]"
            }
        })
    });
}
