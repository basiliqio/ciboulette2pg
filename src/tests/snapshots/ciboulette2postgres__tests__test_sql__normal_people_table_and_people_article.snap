---
source: src/tests/test_sql.rs
expression: "serde_json::Value::Object(map)"

---
{
  "peoples": [
    {
      "id": "[uuid]",
      "first-name": "Francis",
      "last-name": "Le Roy",
      "favorite_color": "[favorite_color_uuid]",
      "age": 22,
      "gender": "M",
      "twitter": null
    },
    {
      "id": "[uuid]",
      "first-name": "Somebody",
      "last-name": "Wuhu",
      "favorite_color": "[favorite_color_uuid]",
      "age": 34,
      "gender": "F",
      "twitter": "@randomhandle"
    },
    {
      "id": "[uuid]",
      "first-name": "AAAAAAAA",
      "last-name": "BBBBBBBBB",
      "favorite_color": "[favorite_color_uuid]",
      "age": null,
      "gender": null,
      "twitter": null
    }
  ],
  "people-article": [
    {
      "id": "[uuid]",
      "people_id": "[uuid]",
      "article_id": "[uuid]"
    },
    {
      "id": "[uuid]",
      "people_id": "[uuid]",
      "article_id": "[uuid]"
    },
    {
      "id": "[uuid]",
      "people_id": "[uuid]",
      "article_id": "[uuid]"
    },
    {
      "id": "[uuid]",
      "people_id": "[uuid]",
      "article_id": "[uuid]"
    }
  ]
}
